### ISSUE TITLE
Roadmap to ConsumerGroup

### ISSUE THREAD

**Issue by DrZhang** [emitter-io/emitter#67]
Concept borrowed from Kafka
Producer published [1,2,3,4,5,6,7,8,9]
ConsumerGroupMember01 subscribes and get [4,5,6]
ConsumerGroupMember02 subscribes and get [2,3]
ConsumerGroupMember03 subscribes and get [1,7,8,9]

**Comment by Kelindar** [comment]
Hi, I assume your question is whether we have a plan to implement consumer groups in emitter, similar to those in Kafka?

**Comment by DrZhang** [comment]
Yes it is.
Sorry for my poor English.

**Comment by Kelindar** [comment]
**tl;dr**: we are indeed considering to have something like this implemented natively in the broker itself, but this is not a priority at the moment.

**Long answer**:  consumer groups are great for poll-based streams/queues mechanisms like Kafka and such re-partitioning of topics is even more important in push-based technologies like emitter. However, as we are push-based we might be better of implementing a load-balancing inside the broker, where a broker might be aware of all of the connected consumers and would like to round-robin messages to consumers which require them. This is not a trivial task and comes with a pre-requisite of supporting at-least-once message delivery semantics, as one or more consumers might be down and will make you loose messages if not consumed. We are definitely are going to work an implement _clustered_ at-least-once delivery, but there's a lot of work on our plate to get there.

This being said, you can view this as a re-partition of a channel and you could do it manually in emitter right now by combining sub-channels and consistent hashing.  For example:

- publisher: _hash(message1) % 3 = 1_ => published to `my-channel/1`
- publisher: _hash(message2) % 3 = 0_ => published to `my-channel/0` 
- publisher: _hash(message3) % 3 = 2_ => published to `my-channel/2`
....

- subscriber 1 subscibes to  `my-channel/0`
- subscriber 2 subscibes to  `my-channel/1`
- subscriber 3 subscibes to  `my-channel/2`

**Comment by DrZhang** [comment]
@Kelindar  
Thanks! 
Sub-channels and consistent hashing is a good idea.
I'll keep watching this project.
Sorry for the impolite way of raising this issues.
